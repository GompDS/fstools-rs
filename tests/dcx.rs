use std::{
    error::Error,
    io::{BufReader, Read},
    path::PathBuf,
};

use crc32fast::Hasher;
use fstools::{formats::dcx::DcxHeader, prelude::*};
use insta::assert_debug_snapshot;

#[test]
pub fn decode_kraken_dcx() -> Result<(), Box<dyn Error>> {
    let er_path = PathBuf::from(std::env::var("ER_PATH").expect("no elden ring path provided"));
    let keys = FileKeyProvider::new("keys");
    let archives = [
        er_path.join("Data0"),
        er_path.join("Data1"),
        er_path.join("Data2"),
        er_path.join("Data3"),
        er_path.join("sd/sd"),
    ];

    let vfs = Vfs::create(archives.clone(), &keys).expect("unable to create vfs");
    let file = vfs.open("/map/m60/m60_44_58_00/m60_44_58_00_445800.mapbnd.dcx")?;
    let (header, mut reader) = DcxHeader::read(file)?;
    let mut hasher = Hasher::new();

    let mut buffer = [0u8; 4096];
    loop {
        match reader.read(&mut buffer) {
            Ok(0) => break, // End of file
            Ok(len) => hasher.update(&buffer[..len]),
            Err(e) => {
                // Handle the error more gracefully, e.g., return it or log it
                eprintln!("Error reading data: {}", e);
                break;
            }
        }
    }

    let hash = hasher.finalize();
    assert_debug_snapshot!((header, hash));

    Ok(())
}
