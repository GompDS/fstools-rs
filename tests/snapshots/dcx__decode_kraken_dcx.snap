---
source: tests/dcx.rs
expression: "(header, hash)"
---
(
    DCX {
        header: Metadata {
            chunk_magic: [
                68,
                67,
                88,
                0,
            ],
            version: U32(
                69632,
            ),
            sizes_offset: U32(
                24,
            ),
            params_offset: U32(
                36,
            ),
            data_info_offset: U32(
                68,
            ),
            data_offset: U32(
                76,
            ),
        },
        sizes: Sizes {
            chunk_magic: [
                68,
                67,
                83,
                0,
            ],
            uncompressed_size: U32(
                2976,
            ),
            compressed_size: U32(
                873,
            ),
        },
        compression_parameters: CompressionParameters {
            chunk_magic: [
                68,
                67,
                80,
                0,
            ],
            algorithm: [
                75,
                82,
                65,
                75,
            ],
            chunk_size: U32(
                32,
            ),
            settings: [
                6,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                1,
                1,
                0,
            ],
        },
    },
    2454203100,
)
